### PR TITLE
UCP/PROTO: Add new protocols for tag offload sync

### PR DESCRIPTION
**Pull request by brminich**
## What
Add tag sync offload protocols with new proto infrastructure

### CODE REVIEW

**Comment by brminich** [comment]
@dmitrygx, @rakhmets can you please review?

File: src/ucp/tag/offload/eager.c [hunk 1]
```diff
@@ -128,13 +125,35 @@ static ucs_status_t ucp_proto_eager_tag_offload_bcopy_init(
...
 
     return ucp_proto_single_init(&params);
 }
 
+
+static ucs_status_t
+ucp_proto_eager_tag_offload_bcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
```
Comment (hoopoepg) [inline src/ucp/tag/offload/eager.c:140]: unaligned '='?

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/tag/offload/eager.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/tag/offload/eager.c:140]: + declare `const` first?

File: src/ucp/tag/offload/eager.c [hunk 2]
```diff
@@ -144,8 +163,42 @@ static ucp_proto_t ucp_eager_bcopy_single_proto = {
 };
 UCP_PROTO_REGISTER(&ucp_eager_bcopy_single_proto);
 
-static ucs_status_t ucp_proto_eager_tag_offload_zcopy_init(
+static ucs_status_t
+ucp_proto_eager_sync_tag_offload_bcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
```
Comment (dmitrygx) [inline src/ucp/tag/offload/eager.c:169]: unaligned by `=`

File: src/ucp/tag/offload/eager.c [hunk 3]
```diff
@@ -201,3 +261,40 @@ static ucp_proto_t ucp_eager_zcopy_single_proto = {
...
+            req, UCT_MD_MEM_ACCESS_LOCAL_READ,
+            ucp_proto_tag_offload_zcopy_sync_send_func, "tag_eager_sync_zcopy");
+}
+
+static ucp_proto_t ucp_eager_sync_zcopy_single_proto = {
+    .name       = "egrsnc/offload/zcopy",
+    .flags      = 0,
+    .init       = ucp_proto_eager_sync_tag_offload_zcopy_init,
+    .config_str = ucp_proto_single_config_str,
+    .progress   = ucp_proto_eager_sync_tag_offload_zcopy_progress,
```
Comment (dmitrygx) [inline src/ucp/tag/offload/eager.c:298]: No need for trailing comma

File: src/ucp/tag/offload/eager.c [hunk 4]
```diff
@@ -144,8 +163,42 @@ static ucp_proto_t ucp_eager_bcopy_single_proto = {
...
+static ucs_status_t
+ucp_proto_eager_sync_tag_offload_bcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    status = ucp_proto_eager_tag_offload_bcopy_common(
+            req, spriv, ucp_send_request_get_ep_remote_id(req));
+    if (status == UCS_OK) {
```
Comment (dmitrygx) [inline src/ucp/tag/offload/eager.c:175]: `ucs_likely`?

**Review by dmitrygx — CHANGES_REQUESTED** [review]

**Review by rakhmets — APPROVED** [review]

File: src/ucp/tag/offload/eager.c (hunk 1 above)
Comment (brminich) [reply src/ucp/tag/offload/eager.c:140]: - regarding alignment, it looks better this way. Imo, looks too ugly if aligned
- why const has to be the first one?

File: src/ucp/tag/eager.h [hunk 5]
```diff
@@ -82,9 +82,9 @@ void ucp_tag_eager_sync_zcopy_completion(uct_completion_t *self);
 
 static UCS_F_ALWAYS_INLINE int
 ucp_proto_eager_check_op_id(const ucp_proto_init_params_t *init_params,
-                            int offload_enabled)
+                            ucp_proto_id_t op_id, int offload_enabled)
```
Comment (yosefe) [inline src/ucp/tag/eager.h:85]: ucp_operation_id_t

File: src/ucp/tag/offload/eager.c [hunk 6]
```diff
@@ -90,25 +91,21 @@ static size_t ucp_eager_tag_offload_pack(void *dest, void *arg)
 }
 
 static ucs_status_t
-ucp_proto_eager_tag_offload_bcopy_progress(uct_pending_req_t *self)
+ucp_proto_eager_tag_offload_bcopy_common(ucp_request_t *req,
```
Comment (yosefe) [inline src/ucp/tag/offload/eager.c:94]: UCS_F_ALWAYS_INLINE ?

File: src/ucp/tag/offload/eager.c [hunk 7]
```diff
@@ -128,13 +125,35 @@ static ucs_status_t ucp_proto_eager_tag_offload_bcopy_init(
...
+static ucs_status_t
+ucp_proto_eager_tag_offload_bcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    status = ucp_proto_eager_tag_offload_bcopy_common(req, spriv, 0ul);
+
+    return ucp_proto_single_status_handle(
```
Comment (yosefe) [inline src/ucp/tag/offload/eager.c:145]: can we move the call to `ucp_proto_single_status_handle` inside `ucp_proto_eager_tag_offload_bcopy_common`

File: src/ucp/tag/offload/eager.c (hunk 1 above)
Comment (yosefe) [reply src/ucp/tag/offload/eager.c:140]: it can't be first because it depends on req

File: src/ucp/tag/offload/eager.c (hunk 7 above)
Comment (brminich) [reply src/ucp/tag/offload/eager.c:145]: we need special handling of operation status (not the one returned by `ucp_proto_single_status_handle`) in `ucp_proto_eager_sync_tag_offload_bcopy_progress`

**Review by yosefe — APPROVED** [review]

**Pull request merged**